### PR TITLE
[Qt] Add out-of-sync modal info layer

### PR DESCRIPTION
**Pull request by jonasschnelli**
Adresses https://github.com/bitcoin/bitcoin/issues/8060 and https://github.com/bitcoin/bitcoin/issues/7235

This change will slide in a semi transparent modal info layer in out-of-sync state resulting in a more prominent warning. Users can hide the modal info layer by pressing on "Hide".
Clicking on the out-of-sync warning icons will re-display the modal info layer.

In the modal info layer, the user can get three new values:
- amount of blocks left
- estimated time left until in-sync
- progress increase per hour 

bildschirmfoto 2016-07-19 um 15 53 16

### CODE REVIEW

**Comment by instagibbs** [comment]
"Progress increase per hour" is a bit weird sounding to me, maybe "Sync progress per hour"?

**Comment by laanwj** [comment]
Looks very nice!

**Comment by laanwj** [comment]
- Looks like the font color should not depend on the theme?
overlay

- Maybe add some more text after "...but this process has not completed yet", to make it more concrete "this means that recent transactions will not be visible, and the balance will not be up-to-date until this process has completed."

**Comment by jonasschnelli** [comment]
Added the second info text and made sure, the font will be blackish (force push).

**Comment by fanquake** [comment]
Testing this, and clicking hide doesn't seem to do anything?
Also, If your opening a close-to, or recently synced wallet, should the modal disappear automatically once it reaches 100%?

OS X screenshots:
modal
full-wallet

Also, this looks like a new compiler warning.
```
  CXX      qt/qt_libbitcoinqt_a-bitcoingui.o
qt/bitcoingui.cpp:118:5: warning: field 'spinnerFrame' will be initialized after field 'modalOverlay' [-Wreorder]
    spinnerFrame(0),
    ^
1 warning generated.
```

**Comment by jonasschnelli** [comment]
@fanquake: It should hide automatically if you are in sync,.. but only, If you haven't opened it manually (by pressing the warning icons).
Not sure if it should also hide in that case, but probably.

**Comment by Joukehofman** [comment]
Look great! Maybe add to the warning text that it's also not possible to send bitcoins?

**Comment by jonasschnelli** [comment]
Fixed warning. Added a short text passage `"Spending bitcoins is not possible during that phase!"`.

**Comment by sipa** [comment]
Does the "This process is not complete yet" disappear when 100% is reached, but the window is still open?

**Comment by jonasschnelli** [comment]
@sipa: yes. It will (https://github.com/bitcoin/bitcoin/pull/8371/files#diff-0db7dd184df07a48c307ccc182021a68R776). But only if you not have manually opened the "window" by clicking on the alert icons (then it will stay intentionally).

**Comment by sipa** [comment]
No, I mean that if you have manually clicked it, and the window does stay
open, does the "This process has not completes yet" text also stay? Because
that is very confusing...

**Comment by jonasschnelli** [comment]
Ah. I get your point. Maybe force closing then would make sense (or changing the text in the upper section [more complicate]).

**Comment by sipa** [comment]
I think force closing is fine.

**Comment by MarcoFalke** [comment]
> I think force closing is fine.


Agree

**Comment by jonasschnelli** [comment]
Updated so that the layer/window does force close when the we detect in-sync state.

**Comment by sipa** [comment]
The "number of blocks left" goes up and down for a while (presumably because the headers aren't processed yet?)... if the tip header is too far in the past, it may be better to say "unknown" there.

**Comment by fanquake** [comment]
Updated OS X screenshot, the hiding issue seems to be fixed now.
modal

**Comment by luke-jr** [comment]
> Looks like the font color should not depend on the theme?

Rather, the background colour *should* depend on the theme *also*? Why is this overriding the theme?

**Comment by flack** [comment]
Shouldn't this dialog also have a headline of some sort (e.g. "Wallet out of sync")?

Because if I look at the screenshot, I only see the progress data, and the text starts with "The displayed information may be out of date", but in fact, it's not the currently displayed data (progress information) that is out of date, but the stuff in the wallet, which is obscured by the overlay.

**Comment by jonasschnelli** [comment]
@luke-jr: I think creating nice visualizations is difficult with the presumption to only use theme based colors.

@flack: Indeed. The term "the displayed information... " can look like its referred to the data below (in the same overlay). I have plans to overhaul this PR soon.

**Comment by jonasschnelli** [comment]
Updated with recommendation from @sipa to hide the remaining blocks during headers-first during IBD.

bildschirmfoto 2016-09-13 um 18 01 34

File: src/qt/modaloverlay.cpp [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+            }
+        }
+        // show progress increase per hour
+        ui->progressIncreasePerH->setText(QString::number(progressPerHour*100, 'f', 2)+"%");
+
+        // show expected remaining time
+        ui->expectedTimeLeft->setText(GUIUtil::formateNiceTimeOffset(remainingMSecs/1000.0));
+
+        // keep maximal 5000 samples
+        static int maxSamples = 5000;
```
Comment (MarcoFalke) [inline src/qt/modaloverlay.cpp]: nit

```cpp
static const int MAX_SAMPLES = 5000;

File: src/qt/modaloverlay.cpp [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+{
+    QDateTime currentDate = QDateTime::currentDateTime();
+
+    // keep a vector of samples of verification progress at height
+    blockProcessTime.push_front(qMakePair(currentDate.currentMSecsSinceEpoch(), nVerificationProgress));
+
+    // show progress speed if we have more then one sample
+    if (blockProcessTime.size() >= 2)
+    {
+        // try to get the window from the last 500 seconds or at least 10 samples
```
Comment (MarcoFalke) [inline src/qt/modaloverlay.cpp]: Where does it say 10 in the source code?

File: src/qt/walletframe.h [hunk 3]
```diff
@@ -38,6 +38,10 @@ class WalletFrame : public QFrame
 
     void showOutOfSyncWarning(bool fShow);
 
+Q_SIGNALS:
+    /** Notify that the user has requested more information about the out-of-sync warning */
+    void requestedOfSyncWarningInfo();
```
Comment (MarcoFalke) [inline src/qt/walletframe.h]: nit: I don't think this is proper English. Idk, get rid of the `Of`?

File: src/qt/bitcoingui.cpp [hunk 4]
```diff
@@ -705,7 +715,17 @@ void BitcoinGUI::setNumConnections(int count)
...
-    if(!clientModel)
+    if (modalOverlay)
+    {
+        if (header)
+        {
+            /* use clientmodels getHeaderTipHeight and getHeaderTipTime because the NotifyHeaderTip signal does not fire when updating the best header */
+            modalOverlay->setKnownBestHeight(clientModel->getHeaderTipHeight(), QDateTime::fromTime_t(clientModel->getHeaderTipTime()));
+        }
+        else
+            modalOverlay->tipUpdate(count, blockDate, nVerificationProgress);
```
Comment (MarcoFalke) [inline src/qt/bitcoingui.cpp]: style nit: brackets are "required" for the else when they are used for the if.

**Review by maflcko — APPROVED** [review]
Tested-only ACK. Found a few style nits. (You can fix them in a separate fixup commit)

ACK d8b062ef5eea3addff00a09bad1dab162452b4b5

**Comment by MarcoFalke** [comment]
Looks good to merge

screenshot from 2016-09-21 10-20-23

**Comment by jonasschnelli** [comment]
Fixed nits reported by @MarcoFalke.
@MarcoFalke: can you re-ack?

**Comment by luke-jr** [comment]
@jonasschnelli On the contrary, violating the theme contradicts the goal of nice visualisations.

**Comment by jonasschnelli** [comment]
@luke-jr: I think adding a black overlay with white text does not directly violates the theme. It just will result in a more controllable look and feel.

**Comment by luke-jr** [comment]
Software controlling look-and-feel means the user's chosen look-and-feel is being ignored. Presumably the user has chosen them because they like it, so violating them simply won't be nice.

**Comment by jonasschnelli** [comment]
I think themes are there for a reason.
Using the theme colors might result in loosing the optical effect which is there for a reason. The back background with the white info layer should imply that the informations within the layer are on a different level of importance.

But happy if someone likes to try migrating this to the theme colors.

**Comment by MarcoFalke** [comment]
re-ACK 080a877

**Comment by luke-jr** [comment]
(BTW, I don't mean to imply theme concerns need to interfere in merging this. By all means, feel free. We can work on theming later.)

**Comment by gmaxwell** [comment]
Too much text, esp too much bolded text. I automatically don't read the top part (though I did in the original version).  Also "Spending bitcoins is not possible during that phase!" is _not_ true.

**Comment by sipa** [comment]
Well you can't spend funds your client hasn't seen the receival of.

**Comment by gmaxwell** [comment]
@sipa Yes, but you _can_ spend anything you can see... This is a common point of confusion already. And in the common case where you had bitcoin off for the last month, but received no payment since then, you're good to go for sending.

**Comment by jonasschnelli** [comment]
@gmaxwell: would you be willing providing a better text?

I agree that "spending Bitcoins is not possible" is not true and can lead to confusion, especially if one catches up just a couple of days (the info-layer will also be shown then).

File: src/qt/forms/modaloverlay.ui [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+             <item>
+              <widget class="QLabel" name="infoTextStrong">
+               <property name="font">
+                <font>
+                 <weight>75</weight>
+                 <bold>true</bold>
+                </font>
+               </property>
+               <property name="text">
+                <string>This means that recent transactions will not be visible, and the balance will not be up-to-date until this process has completed. Spending bitcoins is not possible during that phase!</string>
```
Comment (jonasschnelli) [inline src/qt/forms/modaloverlay.ui]: Or what about [...]Spending bitcoins **may** not be possible[...]?

Comment (MarcoFalke) [inline src/qt/forms/modaloverlay.ui]: Fine, I think either is better.

**Comment by jonasschnelli** [comment]
Force pushed a text- and bold-section-change.
This is how it looks now:
bildschirmfoto 2016-09-23 um 15 31 38

**Comment by MarcoFalke** [comment]
Are you sure you pushed the text change?

**Comment by jonasschnelli** [comment]
Pushed now..

**Comment by MarcoFalke** [comment]
ACK 08827df3ecce925928dc3bedcdef63bfca290300

File: src/qt/clientmodel.cpp [hunk 6]
```diff
@@ -234,7 +250,7 @@ static void BlockTipChanged(ClientModel *clientmodel, bool initialSync, const CB
     int64_t& nLastUpdateNotification = fHeader ? nLastHeaderTipUpdateNotification : nLastBlockTipUpdateNotification;
 
     // if we are in-sync, update the UI regardless of last update time
-    if (!initialSync || now - nLastUpdateNotification > MODEL_UPDATE_DELAY) {
+    if (fHeader || !initialSync || now - nLastUpdateNotification > MODEL_UPDATE_DELAY) {
```
Comment (MarcoFalke) [inline src/qt/clientmodel.cpp]: Why is this needed?

**Comment by Xekyo** [comment]
Instead of "this process" and "that phase" I would repeat "synchronization":

>The displayed information may be out of date.
Your wallet automatically sychronizes with the Bitcoin network while connected, but synchronization has not completed yet. Recent transactions will only become visible as your wallet catches up to the network. The balance will update as transactions are processed.
Spending bitcoins may not be possible until synchronization has finished.
